apm-agent-plugins/apm-jms-plugin/src/main/java/co/elastic/apm/agent/jms/JmsMessageConsumerInstrumentation.java

```diff
@@ -0,0 +1,157 @@
+    }
+
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return named("receive").and(takesArguments(0).or(takesArguments(1))).and(isPublic())
+            .or(named("receiveNoWait").and(takesArguments(0).and(isPublic())));
+    }
+
+    @Override
+    public Class<?> getAdviceClass() {
+        return MessageConsumerAdvice.class;
+    }
+
+    public static class MessageConsumerAdvice {
+
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        @Nullable
+        public static AbstractSpan startSpan(@Advice.Origin Class<?> clazz) {
+            AbstractSpan abstractSpan = null;
+            if (tracer != null) {
+                final TraceContextHolder<?> parent = tracer.getActive();
+                if (parent == null) {
+                    abstractSpan = tracer.startTransaction(TraceContext.asRoot(), null, clazz.getClassLoader())
+                        .withType("messaging");
+                } else {
+                    String parentType = null;
+                    if (parent instanceof Transaction) {
+                        parentType = ((Transaction) parent).getType();
+                    } else if (parent instanceof Span) {
+                        parentType = ((Span) parent).getType();
+                    }
+
+                    // Not creating duplicates for nested receive
+                    if (parentType != null && parentType.equals("messaging")) {
+                        return null;
+                    }
+
+                    abstractSpan = parent.createSpan()
+                        .withType("messaging")
+                        .withSubtype("jms")
+                        .withAction("receive");
+                }
+                abstractSpan.setName("JMS RECEIVE");
+                abstractSpan.activate();
+            }
+            return abstractSpan;
+        }
+
+        @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+        public static void endSpan(@Advice.Enter @Nullable final AbstractSpan abstractSpan,
+                                   @Advice.Return @Nullable final Message message,
+                                   @Advice.Thrown final Throwable throwable) {
+
+            if (abstractSpan != null) {
+                try {
+                    if (message != null) {
+                        if (abstractSpan instanceof Transaction) {
+                            try {
+                                String messageSenderContext = message.getStringProperty(JMS_TRACE_PARENT_HEADER);
+                                if (messageSenderContext != null) {
+                                    abstractSpan.getTraceContext().asChildOf(messageSenderContext);
+                                }
+                            } catch (JMSException e) {
+                                LoggerFactory.getLogger("JmsMessageConsumerInstrumentation")
```

Comment:
I thought it would be misleading to reference a field of the enclosing class, because in runtime, it is not really the enclosing class, just a random agent class...
However, using the `VisibleForAdvice` may be enough to indicate that. 
Will do.